File: python/ch17_agent/agent_ch17/core/profiles.py
"""固定章节能力快照。

Java 对照：`ChapterProfile` 类似不可变配置 record，P01-P06 类似预定义单例常量。
组合根使用对象身份判断，调用方不能临时拼一个同字段对象来冒充正式章节。
"""

from dataclasses import dataclass
from typing import Literal

Capability = Literal[
    "loop",
    "powershell",
    "tool_registry",
    "files",
    "policy",
    "hooks",
    "todo",
    "subagent",
    "skills",
    "artifacts",
    "compaction",
    "memory",
    "dynamic_prompt",
    "recovery",
    "task_dag_json",
    "background",
    "cron",
    "teammate",
    "mailbox",
    "protocol",
    "plan_gate",
    "task_dag_sqlite",
    "work_stealing",
]


@dataclass(frozen=True, slots=True)
class ChapterProfile:
    """章节号及其允许装配的能力白名单。"""

    chapter: int
    capabilities: frozenset[Capability]


P01 = ChapterProfile(1, frozenset({"loop", "powershell"}))
P02 = ChapterProfile(2, frozenset({"loop", "powershell", "tool_registry", "files"}))
P03 = ChapterProfile(3, frozenset({"loop", "powershell", "tool_registry", "files", "policy"}))
P04 = ChapterProfile(
    4, frozenset({"loop", "powershell", "tool_registry", "files", "policy", "hooks"})
)
P05 = ChapterProfile(
    5, frozenset({"loop", "powershell", "tool_registry", "files", "policy", "hooks", "todo"})
)
P06 = ChapterProfile(
    6,
    frozenset(
        {"loop", "powershell", "tool_registry", "files", "policy", "hooks", "todo", "subagent"}
    ),
)
P07 = ChapterProfile(
    7,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
        }
    ),
)

P08 = ChapterProfile(
    8,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
            "artifacts",
            "compaction",
        }
    ),
)

P09 = ChapterProfile(
    9,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
            "artifacts",
            "compaction",
            "memory",
        }
    ),
)

P10 = ChapterProfile(
    10,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
            "artifacts",
            "compaction",
            "memory",
            "dynamic_prompt",
        }
    ),
)

P11 = ChapterProfile(
    11,
    frozenset(
        {
            "loop",
            "powershell",
            "tool_registry",
            "files",
            "policy",
            "hooks",
            "todo",
            "subagent",
            "skills",
            "artifacts",
            "compaction",
            "memory",
            "dynamic_prompt",
            "recovery",
        }
    ),
)

P12 = ChapterProfile(12, frozenset((*P11.capabilities, "task_dag_json")))
P13 = ChapterProfile(13, frozenset((*P12.capabilities, "background")))
P14 = ChapterProfile(14, frozenset((*P13.capabilities, "cron")))
P15 = ChapterProfile(15, frozenset((*P14.capabilities, "teammate", "mailbox")))
P16 = ChapterProfile(16, frozenset((*P15.capabilities, "protocol", "plan_gate")))
P17 = ChapterProfile(17, frozenset((*P16.capabilities, "task_dag_sqlite", "work_stealing")))


def profile_for_chapter(chapter: int) -> ChapterProfile:
    """只返回模块内固定单例；尚未迁移的章节明确报错。"""
    profiles = {
        1: P01,
        2: P02,
        3: P03,
        4: P04,
        5: P05,
        6: P06,
        7: P07,
        8: P08,
        9: P09,
        10: P10,
        11: P11,
        12: P12,
        13: P13,
        14: P14,
        15: P15,
        16: P16,
        17: P17,
    }
    if isinstance(chapter, bool) or not isinstance(chapter, int) or not 1 <= chapter <= 20:
        raise ValueError("chapter 必须是 1 到 20 的整数")
    if chapter not in profiles:
        raise ValueError(f"第 {chapter} 章尚未迁移为 Python")
    return profiles[chapter]

File: typescript/chapters/ch20/src/core/profiles.ts
// 章节能力清单：每个 profile 都是冻结单例，组合根据此决定启用的运行时边界。
export type Capability =
  | "loop"
  | "powershell"
  | "tool_registry"
  | "files"
  | "policy"
  | "hooks"
  | "todo"
  | "subagent"
  | "skills"
  | "artifacts"
  | "compaction"
  | "memory"
  | "dynamic_prompt"
  | "recovery"
  | "task_dag_json"
  | "background"
  | "cron"
  | "teammate"
  | "mailbox"
  // protocol 与 plan_gate 共同表示结构化审批消息和副作用执行门控。
  | "protocol"
  | "plan_gate"
  // task_dag_sqlite 与 work_stealing 表示带租约的 SQLite 任务图与去中心化认领。
  | "task_dag_sqlite"
  | "work_stealing"
  // worktree 表示受控 Git 工作树隔离，mcp 表示可动态发布和撤销的远程 MCP 工具边界。
  | "worktree"
  | "mcp"
  // full_harness 标记 P20 完整能力集，不新增独立运行时。
  | "full_harness";

export interface ChapterProfile {
  readonly chapter: number;
  readonly capabilities: ReadonlySet<Capability>;
}

class CapabilitySet implements ReadonlySet<Capability> {
  readonly #values: Set<Capability>;

  constructor(values: readonly Capability[]) {
    this.#values = new Set(values);
  }

  get size(): number {
    return this.#values.size;
  }

  has(value: Capability): boolean {
    return this.#values.has(value);
  }

  entries(): SetIterator<[Capability, Capability]> {
    return this.#values.entries();
  }

  keys(): SetIterator<Capability> {
    return this.#values.keys();
  }

  values(): SetIterator<Capability> {
    return this.#values.values();
  }

  forEach(
    callbackfn: (value: Capability, value2: Capability, set: ReadonlySet<Capability>) => void,
    thisArg?: unknown,
  ): void {
    this.#values.forEach((value) => {
      callbackfn.call(thisArg, value, value, this);
    });
  }

  [Symbol.iterator](): SetIterator<Capability> {
    return this.values();
  }
}

// 每章只声明新增能力；后续章节通过累计 PROFILE_DELTAS 得到完整档案。
const PROFILE_DELTAS = [
  ["loop", "powershell"],
  ["tool_registry", "files"],
  ["policy"],
  ["hooks"],
  ["todo"],
  ["subagent"],
  ["skills"],
  ["artifacts", "compaction"],
  ["memory"],
  ["dynamic_prompt"],
  ["recovery"],
  ["task_dag_json"],
  ["background"],
  ["cron"],
  ["teammate", "mailbox"],
  ["protocol", "plan_gate"],
  ["task_dag_sqlite", "work_stealing"],
  ["worktree"],
  ["mcp"],
  // full_harness 是 P20 的标记能力：不新增具体运行时，只声明前十九章能力全部生效。
  ["full_harness"],
] as const satisfies readonly (readonly Capability[])[];

// PROFILES 由增量能力数组推导，保证每章档案都是前一章的严格超集。
const PROFILES: readonly ChapterProfile[] = Object.freeze(
  PROFILE_DELTAS.map((_, index) =>
    Object.freeze({
      chapter: index + 1,
      capabilities: new CapabilitySet(PROFILE_DELTAS.slice(0, index + 1).flat()),
    }),
  ),
);

export const P01 = profileAt(1);
export const P02 = profileAt(2);
export const P03 = profileAt(3);
export const P04 = profileAt(4);
export const P05 = profileAt(5);
export const P06 = profileAt(6);
export const P07 = profileAt(7);
export const P08 = profileAt(8);
export const P09 = profileAt(9);
export const P10 = profileAt(10);
export const P11 = profileAt(11);
export const P12 = profileAt(12);
export const P13 = profileAt(13);
export const P14 = profileAt(14);
export const P15 = profileAt(15);
export const P16 = profileAt(16);
export const P17 = profileAt(17);
export const P18 = profileAt(18);
export const P19 = profileAt(19);
// P20 是完整 Harness 档案，统一启用前十九章已经验证的累计能力。
export const P20 = profileAt(20);

export function profileForChapter(chapter: number): ChapterProfile {
  // 只有 1 到 20 的固定章节可查询；越界立即失败，避免调用方构造未知档案。
  if (!Number.isInteger(chapter) || chapter < 1 || chapter > PROFILES.length) {
    throw new Error("chapter must be an integer from 1 to 20");
  }
  return profileAt(chapter);
}

function profileAt(chapter: number): ChapterProfile {
  const profile = PROFILES[chapter - 1];
  // 未迁移章节在运行前显式失败，而不是返回空能力集。
  if (profile === undefined) {
    throw new Error(`Chapter ${chapter} has not been migrated to TypeScript yet`);
  }
  return profile;
}
